catalog/registry/catalog-registry-api-impl/src/main/java/ddf/catalog/registry/api/impl/RegistryStoreImpl.java

```diff
@@ -0,0 +1,235 @@
+        }
+        return super.create(request);
+    }
+
+    @Override
+    public UpdateResponse update(UpdateRequest request) throws IngestException {
+
+        if (!pushAllowed) {
+            throw new IngestException("Push (write) operations are not allowed for this registry");
+        }
+
+        Map<String, Metacard> updatedMetacards = new HashMap<>();
+        Map<String, Metacard> origMetacards = new HashMap<>();
+
+        for (Map.Entry<Serializable, Metacard> entry : request.getUpdates()) {
+
+            updatedMetacards.put(getRegistryId(entry.getValue()), entry.getValue());
+        }
+
+        OperationTransaction opTrans =
+                (OperationTransaction) request.getPropertyValue(Constants.OPERATION_TRANSACTION_KEY);
+
+        for (Metacard mcard : opTrans.getPreviousStateMetacards()) {
+            origMetacards.put(getRegistryId(mcard), mcard);
+        }
+
+        //update the new metacards with the id from the orig so that they can be found on the remote system
+        try {
+            for (Map.Entry<String, Metacard> entry : updatedMetacards.entrySet()) {
+                setMetacardExtID(entry.getValue(),
+                        origMetacards.get(entry.getKey())
+                                .getId());
+            }
+        } catch (ParserException | IOException e) {
+            throw new IngestException("Could not update metacards id", e);
+        }
+
+        return super.update(request);
+    }
+
+    @Override
+    public DeleteResponse delete(DeleteRequest request) throws IngestException {
+        if (!pushAllowed) {
+            throw new IngestException("Push (write) operations are not allowed for this registry");
+        }
+        return super.delete(request);
+    }
+
+    @Override
+    public SourceResponse query(QueryRequest request) throws UnsupportedQueryException {
+        if (!pullAllowed) {
+            throw new UnsupportedQueryException(
+                    "Pull (read) operations are not allowed for this registry");
+        }
+        return super.query(request);
+    }
+
+    public void setPushAllowed(boolean pushAllowed) {
+        this.pushAllowed = pushAllowed;
+    }
+
+    public void setPullAllowed(boolean pullAllowed) {
+        this.pullAllowed = pullAllowed;
+    }
+
+    private String getRegistryId(Metacard mcard) {
+        Attribute attr = mcard.getAttribute(RegistryObjectMetacardType.REGISTRY_ID);
+        if (attr != null && attr.getValue() instanceof String) {
+            return (String) attr.getValue();
+        }
+        return null;
+    }
+
+    private void setMetacardExtID(Metacard metacard, String newId)
+            throws ParserException, IOException {
+
+        String metadata = metacard.getMetadata();
+
+        InputStream inputStream = new ByteArrayInputStream(metadata.getBytes(Charsets.UTF_8));
+        FileBackedOutputStream outputStream = new FileBackedOutputStream(1000000);
```

Comment:
Why did you need to use a FBOS here? You're already loading the entire thing into memory above anyways. You should also put these into a try with resources so they can be closed (you're never closing the FBOS).